### DISCUSSION TITLE
How to control a Multi 4in1 module with a PC?

### DISCUSSION THREAD

**Discussion by Ralung** [pascallanger/DIY-Multiprotocol-TX-Module#501] [Q&A]
Hello, 

#### I am trying to modify the _config.h file to let my 4 in 1 board start auto binding while I plug it in my laptop or after I upload the code with Arduino IDE.
But I couldn't find a correct place and I am not sure how many file I need to modify to implement the function.
Term | My situation
------|------
Multiprotocol code version | v1.3.1.92 (compiling and updating firmware using Arduino IDE)
4 in 1 board | JP 4 in 1 v1.3(c)
Mode | Serial mode
LED status | Slow blink(on=0.5s,off=0.5s) while 4 in 1 board is plugged in USB port of laptop
Drone | E016H
Protocol | E016HV2
Transmitter | RadioMaster TX16s

I have added statement, "AUTOBIND_FLAG_on", in the line 1800 in Multiprotocol.ino file and comment the statement, "#define WAIT_FOR_BIND", in the line 66 in _config.h file, but these didn't work.

The 4 in 1 board was tested to work well before I modified the files. I can use it to fly my drone and it will also start auto binding once I use the "bind on channel" function.

#### Can anyone give me some suggest about how to modify the _config.h file?
If there are anything unclear, please let me know.

**Comment by pascallanger** [comment]
You are saying that you are using serial mode to control module. Then just set the bind flag in the serial frame sent to the module.

**Reply by Ralung** [reply]
Do you mean to send the command from the serial monitor in Arduino IDE?
![image](https://user-images.githubusercontent.com/76200428/104398282-64812d80-5589-11eb-9ff3-913720f288f3.png)

#### I couldn't send the commands in Serail/ FTDI Debugging mode, but I got some data from the serial frame.

![fine_tuning_+2](https://user-images.githubusercontent.com/76200428/104419424-6b249a80-55b3-11eb-8d51-62b67720e5e3.JPG)

The following statement showed up when I press bind button or pull down the rebound lever which is the shortcut I set in channel 16 for binding.

> P: A0 80 4A 94 4C 4C 4C 4C 00 80 00 F6FC

I don't understand the meaning of each number or letter, but I have tried to send this statement on the top of serial monitor and it didn't work.

#### What is the correct method to send the serial commands to the 4 in 1 board?

I also found one place that I am confused.  I have set the "RF Freq. fine tune" as +20, but it seems to be changed back to +2?

![fine_tuning_+20](https://user-images.githubusercontent.com/76200428/104421192-fc950c00-55b5-11eb-9e4f-7cfd275a7611.JPG)

BTW, why does the serial monitor keep blank when I plug the 4 in 1 board directly in the laptop?

**Comment by kilrah** [comment]
The 2nd byte of the control frame you send to the multi contains the bind bit, if that's 1 the module will go in bind mode.

https://github.com/pascallanger/DIY-Multiprotocol-TX-Module/blob/b516bb8d20989ff920f61e1126954bfe28b0b9a1/Multiprotocol/Multiprotocol.h#L770

If you still haven't figured that out you send that whole 27-36 byte control frame to the module continuously.

**Reply by Ralung** [reply]
Because I can't send anything in my serial monitor.
I first uncomment "MULTI protocol serial definition", which is from line 757 to line 1077 in Multiprotocol.h file.
Then I got errors like:

![image](https://user-images.githubusercontent.com/76200428/104430407-96ae8180-55c1-11eb-915a-5d8eba3aef76.png)

Could you please give me an example about sending commands?
Thank you!

**Reply by kilrah** [reply]
> Because I can't send anything in my serial monitor.

If you use realterm you can send frames in hex format.

But you'll need to write your own software to control the module as it's designed to anyway, a serial moitor is of no use. 

> I first uncomment "MULTI protocol serial definition", which is from line 757 to line 1077 in Multiprotocol.h file.
>Then I got errors like:

Why would you do that? It's not code, it's the documentation of the protocol that's there for you to read to know what you need to send to the module.

**Reply by Ralung** [reply]
Thank you, Mr. Kilrah.

I will search how to use Realterm first, but what is the meaning of "frames"?  Is that a kind of commands?
I know how to send commands to Arduino serial port with Python and I will try it after I read the documentation.

Do I need to give another power support when I plug the 4 in 1 board in my laptop with USB port?
Because when I use the USB port of the computer to power the 4 in 1 module, the way the red LED flashes is different from when I use the five pin in the back of my TX16s to power the module.

This is my 4 in 1 module.
![JP_4in1_v1 3(c)](https://user-images.githubusercontent.com/76200428/104436260-5999bd80-55c8-11eb-8513-46688c292548.jpg)

**Reply by kilrah** [reply]
1. You can't control the module through the usb port, you need to give a serial stream to the top pin of the module connector like a radio would using a separate usb-serial adapter (unless there's a debug option available that I don't know of).

2.  A data frame is a stream of bytes that correspond to the protocol definition. E.g. 
`0x55 0x01 0x00 0x00 0x00 0x04 0x20 0x87 0x07 0x08 0x40 0x00 0x02 0x10 0x80 0x00 0x04 0x20 0x00 0x01 0x08 0x40 0x00 0x02 0x10 0x80 0x08`
sent as raw binary is a valid MP frame, and that's sent periodically, e.g. 100 times per second with the content you need.

See the documentation, 
0x55 = "sub_protocol values are 0..31	Stream contains channels"
0x01 = sub_protocol = Flysky, no bind, no range check and no autobind  

etc...

**Reply by Ralung** [reply]
Thank you for telling me so many information!

For the usb-serial adapter, do I need the [3.3v USB-TTL](https://github.com/pascallanger/DIY-Multiprotocol-TX-Module/blob/master/docs/Compiling_STM32.md#tools-required) or the [3.3V USBasp Progranner](https://github.com/pascallanger/DIY-Multiprotocol-TX-Module/blob/master/docs/Compiling.md#tools-required)?

Or just a usb adapter which contains Rx pin, Tx pin, 3.3V pin and GND pin?

I also noticed one thing that is you said "**give a serial stream to the top pin of module**", but the guy who sold me the 4 in 1 board told me that the top pin is for PPM mode. 
There is the picture he gave me.
![pin_TX16s](https://user-images.githubusercontent.com/76200428/104443331-450df300-55d1-11eb-902a-3a002881e5cf.jpg)

It make me confused that what kind of adapter should I use?
If it is convenient, could you please share me a link?

**Reply by kilrah** [reply]
The top pin is PPM if the module is in PPM mode, and serial if the module is in serial mode.

**Comment by pascallanger** [comment]
The module doesn't do anything if it is not receiving a serial stream.
If you switch the module to PPM using the dial and configure the banks in config.h to use the correct protocol then it might send a bind but it will stop shortly if no ppm signal is available.

Anyway what are you trying to achieve?
I have the impression that you want to control the module from your PC?

**Comment by Ralung** [comment]
> The module doesn't do anything if it is not receiving a serial stream.

Does this mean I have to send the command in correct format?
But I tried to send "**P: A0 80 4A 94 4C 4C 4C 4C 00 80 00 F6FC**" but the module didn't start to bind.

There is no rotary switch and no binding button on my 4 in 1 module, so I tried with serial mode first.
(It is difficult to buy a 4 in 1 board with rotary switch and binding button in where I live.)

Yes, Mr. Pascal.
I would like to control three E016H from my PC and the first step is to control one E016H from my PC.
That is another reason that I prefer to choose serial mode which can send the commands but not the PPM mode which needs a binding button.

If I misunderstand anything, please let me know.
Thank you!

**Reply by pascallanger** [reply]
By the way, PPM does not need a "binding button"... This is an auto bind protocol meaning that as soon as the protocol is selected (by using the trigger dial position which is just a short between 2 pins) it will send a bind. But anyway...

**Reply by Ralung** [reply]
Thank you!

**Comment by pascallanger** [comment]
@goebish you've done something similar for one of your friend. Is there a possibility to share your work? If I remember in your modified version you are sending the commands by USB and have put a work around in place for the bootloader.

**Reply by goebish** [reply]
The workaround to "skip" the bootloader when starting the module from usb is to run stm32flash, i.e:
stm32flash -g 0x8002000 COM3
(thanks @benlye ;))

**Reply by gvalbusa** [reply]
@goebish is the code still available? I could not find the file tx_example.py @Ralung mentioned in the next post in your repository. 

Thanks for all the code shared

**Reply by goebish** [reply]
it's on this branch:
https://github.com/goebish/DIY-Multiprotocol-TX-Module/tree/JP4-in-1_USB_Serial_input

**Reply by gvalbusa** [reply]
Thanks!

**Comment by kilrah** [comment]
If I understand well it's a school project and figuring it out/writing the associated code or at least finding it themselves is kind of the point of the assignment...

**Reply by goebish** [reply]
Sorry, I replied to Pascal without reading the thread before ... that's just a crude sample with pretty much everything hardcoded for a specific usage (control 4x RC boats with a single TX module), some work is required for "real world" usage.

**Reply by Ralung** [reply]
Yes, Mr. Kilrah, this is my personal school project.
I am sorry that I asked some trivial questions.
Thank you very much again for your guidance!

**Comment by Ralung** [comment]
Hello, Mr. @goebish 

I have read your tx_example.py and tried to add some codes for my situation.

These are the codes I added
> E016VH2 = 80
> protocol = E016HV2
> AutoBindBit = 0x40
> RangeCheckBit = 0x20
> header = bytearray([0x55, protocol | BIND_FLAG | RangeCheckBit | AutoBindBit, 0x00, 0x00])
> serial_handle.write(bytearray([0x55, 0x01, 0x00, 0x01, 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]))

and there are the outputs I got
>starting JP4in1 firmware
stm32flash.exe -g 0x8002000 COM5
Multi-Module is running
opening COM5 @ 50000 8E2
module waiting for commands
sending bind packets for 10 seconds
starting channels transmition
sweeping rudders, press CTRL+C to stop ...
Traceback (most recent call last):
  File "C:\Users\User\Desktop\tx_example\tx_example_test1.py", line 204, in <module>
    time.sleep(0.02)
KeyboardInterrupt

The outputs said "Multi-Module is running" when I run the program right after I just plug in my 4 in 1 module.
But the Red LED didn't flash and it was always not bright.
I also tried to modify the header as "header = bytearray([0x55, protocol | BIND_FLAG | RangeCheckBit | AutoBindBit])", but it didn't work either.

Could you please share some experiences about how to deal with this issue?
Thank you

**Reply by goebish** [reply]
I'm not sure what the problem is (the led should blink after the stm32flash command) but for this protocol you don't have to use the BIND_FLAG nor AUTOBIND, this protocol always sends bind packets during startup.
How are you sending commands to the module ? directly via USB or using an USB to UART adapter connected to the JR connector on the module ?
If this is directly via USB a modified firmware is required to take input from UART1.

**Reply by Ralung** [reply]
I send the commands directly via USB, and I flashed the firmware by Arduino IDE and used Serial/FTDI Debugging mode.

Maybe STM32FLASH command isn't set properly for my laptop. 
Because I can't open stm32flash.exe with double click.

If the BIND_FLAG and AUTOBIND are not need, should I modify the commands like this?
> header = bytearray([0x55, protocol, 0x00, 0x00])
serial_handle.write(bytearray([0x55, 0x00, 0x00, 0x00, 0x00,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]))

Sorry, I have read the MULTI protocol serial definition in Multiprotocol.h file, but I am still confused about how to construct a correct command...

**Reply by goebish** [reply]
stm32flash runs fine on your machine.
Try disabling Debugging since the uart connected to usb is used for commands now, it can't be used for debugging (at least not without some work)

**Reply by goebish** [reply]
and yes, header = bytearray([0x55, protocol, 0x00, 0x00]) should be fine for the E016Hv2, no need for | BIND_FLAG or anything else

edit: I'm wrong, read below

**Reply by goebish** [reply]
I think I understand why it doesn't work.
The git branch with the USB uart "patch" is dated from last September, E016HV2 didn't exist back then.
Look at the required changes in Multiprotocol.ino and Pins.h to use uart1 instead of uart2 then patch latest Pascal's branch with that
https://github.com/pascallanger/DIY-Multiprotocol-TX-Module/compare/master...goebish:JP4-in-1_USB_Serial_input

... or just use git to merge my branch into the latest version, it should be able to merge automatically (without conflicts), I let you read git documentation if you want to take that route ;)

**Reply by goebish** [reply]
Also, I see this protocol has number 80, so a different header is required as well (sorry)
`header = bytearray([0x54, protocol + 32, 0x00, 0x00])`
because the protocol number is above 31

**Comment by Ralung** [comment]
Hello, Mr. @goebish 

Thank you for helping me, and I am sorry for replying you til today.
I was in a meeting yesterday and I just have time today to test the codes.

I have modified the multiprotocol.ino file and pins.h file.
But I got these errors
>C:\Users\User\AppData\Local\Temp\arduino_build_128608/..\arduino_cache_493587\core\core_3a37eeec5f7896aa26ade432961a153a.a(usart_f1.c.o): In function `__irq_usart1':
C:\Users\User\Documents\ArduinoData\packages\multi4in1\hardware\STM32F1\1.2.2\system/libmaple/usart_private.h:46: multiple definition of `__irq_usart1'
sketch\Multiprotocol.ino.cpp.o:D:\multi_protocol\test\DIY-Multiprotocol-TX-Module-master-test1\Multiprotocol/Multiprotocol.ino:2492: first defined here
collect2.exe: error: ld returned 1 exit status
exit status 1

I am trying to fix these errors now, and the good news is that the Red LED flashed after using the statements as you said.
>header = bytearray([0x54, protocol + 32, 0x00, 0x00])
serial_handle.write(bytearray([0x54, 0x01, 0x00, 0x00, 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]))

But the mode of Red LED is the "Slow blink" mode not the "Fast blink" mode which the mode showed up when I pressed the bind button on my transmitter.
> Slow blink(on=0.5s,off=0.5s): serial has been selected but no valid signal is being seen on the RX pin.
Fast blink(on=0.1s,off=0.1s): bind in progress

I am testing with different commands and will update if I find anything new.
Thank you!

**Reply by goebish** [reply]
regarding the compilation error, I'm not sure that's the right way to do but as the commit message says I had to comment out the __irq_usart1 function in C:\Users\goebish\AppData\Local\Arduino15\packages\multi4in1\hardware\STM32F1\1.1.8\cores\maple\libmaple\usart_f1.c (near the end of the file, Interrupt handlers)

**Reply by goebish** [reply]
The command would be: 
```
protocol = 80  # E016Hv2
header = bytearray([0x54, protocol + 32, 0x00, 0x00])
# send header + channels values
serial_handle.write(header + bytearray([0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]))
```
(not sure about the 0x54 and +32 for E016Hv2, @pascallanger can you confirm ? Multiprotocol.h is not clear and I'm too lazy to look at the OpenTX source code)

**Reply by goebish** [reply]
since I'm not sure for the 0x54 and +32 in the header you can use a dirty trick:
in the PROTOCOLS enum in Multiprotocol.h, swap PROTO_E016HV2 with another one that has a number below 32, e.g.
```
// swap Q303 & E016HV2 protocol number
PROTO_Q303 = 80
...
PROTO_E016HV2 = 31
```
Then the required header will be [0x55, 31, 0x00, 0x00]
That's not the proper way to do! But at least I'm pretty confident that it should work ;)

**Reply by pascallanger** [reply]
There is no +32 since the protocol is sent as a bitmask of 0x1F. To access protocols above 63, you need to send an extra byte at the end of the frame which contains the 2 top bits and concatenated with the header 2 bits extend the protocols to 256.
I know it's not simple but that's how the project has evolved over time. Who could have guessed at the begining when only a few (single digit) number of protocols were available would get so large.
But as you said the easiest is to remap the protocol to less than 32.

**Comment by Ralung** [comment]
Thank you Mr. @goebish  and Mr. @pascallanger 

After I comment out the __irq_usart1 function in usart_f1.c file at
> C:\Users\User\Documents\ArduinoData\packages\multi4in1\hardware\STM32F1\1.2.2\cores\maple\libmaple

There are no error when compiling.

Then, I modified the protocol number of  E016HV2 in multiprotocol.h file as 31
> PROTO_E016HV2 = 31

, and I modified the header as below
> header = bytearray([0x55, 31, 0x00, 0x00])

After that, the 4 in 1 board can turn to the binding process successfully (the Red LED is in "Fast blink" mode), and LEDs of my E016H will stop blinking. I thought that means the binding process completed.
But the Red LED of 4 in 1 board would not stop blinking after the LEDs of my E016H are stop blinking. The binding process will be continued and the Red LED of 4 in 1 board can't keep turn on stably. Which resulted in, during the process of "send_bind_packets" function, the E016H will switch from "bind" mode and "unbind" mode two or three times.

According to above problem, I modified the header as 
> header = bytearray([0x55, protocol, 0x00, 0x01])
or header = bytearray([0x55, protocol, 0x00, option_protocol]) (opetion_protocol = 40)

Because E016HV2 protocol needs Freq. fine tuning.

After set a suitable value of Freq. fine tuning, the Red LED of 4 in 1 module will keep turn on stably as well as LEDs of my E016H.
But when the "send_bind_packets" is end, the RED LED will turn to "slow blink" mode again.
I think it might be because I couldn’t make the propeller turn successfully. I have tried to modify the numbers of channel as well as send different values to different channels, but the problems are still not solved. 

I am trying to solve that and will update if I find any solution. Thank you!

**Reply by goebish** [reply]
Nice, that's some progress already!
I'll let you search now, I don't have an E016H anymore anyway so I can't test ;)

**Reply by pascallanger** [reply]
Your problem is that the module needs to receive serial data all the time. If it does not receive anything for a period of time it will stop sending and slow blink.
That has nothing to do with throttle or anything else like this...

**Reply by goebish** [reply]
That's stated in the python sample code, serial frames must be sent to the module at least once every 70ms (this can be changed in the firmware, but don't).

**Reply by Ralung** [reply]
Mr. @goebish and Mr.@pascallanger , thank you both very very much for so much information!
The propeller is turning successfully now !!! 
I am really appreciate for your help! 👍 =)))

**Reply by dataphile-jeff** [reply]
Would you be willing to share the final code from your project?  I know there are many people who are looking to control drones from their PC for personal projects.  In particular, could you share the tx_example.py that @goebish sent you?  Also, did you end up using an Arduino to sit between your computer and the 4-in-1 board?  

Any help you can provide is much appreciated!

**Reply by goebish** [reply]
tx_example.py and modified firmware for the JP 4-in-1 (~1 year old, not updated) to take serial input from USB instead of JR connector is available here  :
https://github.com/goebish/DIY-Multiprotocol-TX-Module/tree/JP4-in-1_USB_Serial_input
An Arduino board is not required.

**Reply by dataphile-jeff** [reply]
> tx_example.py and modified firmware for the JP 4-in-1 (~1 year old, not updated) to take serial input from USB instead of JR connector is available here :
> https://github.com/goebish/DIY-Multiprotocol-TX-Module/tree/JP4-in-1_USB_Serial_input
> An Arduino board is not required.

Really appreciate it!

**Reply by Ralung** [reply]
@dataphile-jeff I control the drones with just computer and 4-in-1 board in the end. Let me organize the codes and documents and I would share to you😊.

**Reply by goebish** [reply]
@Ralung please share publicly if you can 😊

**Reply by Ralung** [reply]
@dataphile-jeff  @goebish  I just finished my first article: 
https://github.com/Ralung/Control-and-fly-a-mini-drone-with-computer-keyboard
Please let me know if there any problems or unclear part.

**Reply by dataphile-jeff** [reply]
> @dataphile-jeff @goebish I just finished my first article:
> https://github.com/Ralung/Control-and-fly-a-mini-drone-with-computer-keyboard
> Please let me know if there any problems or unclear part.

Thanks!  From the volume of questions that @pascallanger seems to get on the subject of PC drone control, this seems like it will be helpful to a lot of people.

**Reply by wvthoog** [reply]
@Ralung

Awesome ! Thank you so much

**Reply by dataphile-jeff** [reply]
@goebish @pascallanger This is probably a long shot, but ... 

I've been trying to fly a Cheerson CX-10 using an NRF24l01+ module connected to an Arduino UNO board, following [this example](https://makehardware.com/2016/04/24/teach-your-pc-to-fly-a-mini-drone/). The example is based on older code by @goebish, and is modified to send serial commands in a python script. However, when Ralung got a solution to work (and with fewer components) I decided to abandon ship and switch to the 4-in-1 module and an E016H.

However, it seems my JP4-in-1 from Banggood is on backorder forever (as it is everywhere else) and so I was thinking of trying one more go at the Arduino/NRF24l01+/Cheerson.  I'm somewhat certain that the issue in the code is with binding to the Cheerson.  With limited knowledge of drones, I can't understand exactly what's going wrong, but by comparing back and forth against the DIY-Multiprotocol-TX-Module protocol (CX10_NRF24L01) I can see that the binding procedure is fairly similar.  There's the same reference to setting packet[5]-packet[8] during the binding procedure, and both seem to listening for packet[9] =1. 

I've been pouring over the code, but my lack of knowledge about drones is inhibiting me from understanding if the old code from @goebish fully matches the latest protocol in DIY-Multiprotocol-TX-Module.  If you are willing to provide some insight, it would be greatly appreciated (it might be nice to show two use cases for drone control by PC).  If not, then I understand you are busy, and I'm going to follow the path from Ralung when my 4-in-1 module eventually arrives.

**Reply by pascallanger** [reply]
I've already answered you on RCGroups that the CX-10 has a large number of variants and there might be a new one put there. The code you describe is for the blue PCB version (that's how at that time it was recognizable). Have you tried all the variants=sub protocols already (12 in total)?
Even the E016H has 2 different protocols (for now?)...
So basically we don't know if your hardware is working or not, we don't know which sub protocol the cx-10 is using and even if it is a known sub protocol. That's a lot of unknowns...

**Reply by dataphile-jeff** [reply]
Appreciate the response both here and on RCGroups.  I looked into the subvariants, but I didn't see in the code how they would be relevant (only the green and blue PCB boards were specified in the goebish code, and the alternatives seemed to mention picture and video, which the drone does not have).  It is probably that I am quite new to this, and am not clear on how the other protocols would be different.  Would it be helpful to me to take out the four screws to try and look at the board?

If I can't make the Cheerson work, I'm just crossing my fingers that a literal copy of Ralung's procedure will work.  

The last thing I want to do is be annoying ... if there's not an easy way to provide guidance, then I totally understand that this is not the direct mission of your project.  It's unfortunate that there isn't a cheap drone intentionally meant for PC control, as it seems there is a demand for such an outcome.

**Reply by pascallanger** [reply]
The problem right now is that I'm not even sure your hardware works. You say that you are not even running the multi code on the arduino uno, therefore I'm not sure what to do. At least the multi code running on the arduino uno would tell you if the PC communication works, what it is doing like bind or normal. Right now since you are not even using the code from this github how do you want to get support remotely on your setup?
Again, there are many models of the CX-10 with different PCB but also the CX10D, CX10WD...
If you really want to know if your version of CX-10 is compatible turn your nrf24l01 into a RX and listen to the XN297 packets from the original TX over the air. That would tell you if your nrf24l01 is working as it should, that would also tell you in the end if the protocol is known or not.

**Reply by dataphile-jeff** [reply]
> The problem right now is that I'm not even sure your hardware works. You say that you are not even running the multi code on the arduino uno, therefore I'm not sure what to do. At least the multi code running on the arduino uno would tell you if the PC communication works, what it is doing like bind or normal. Right now since you are not even using the code from this github how do you want to get support remotely on your setup?
> Again, there are many models of the CX-10 with different PCB but also the CX10D, CX10WD...
> If you really want to know if your version of CX-10 is compatible turn your nrf24l01 into a RX and listen to the XN297 packets from the original TX over the air. That would tell you if your nrf24l01 is working as it should, that would also tell you in the end if the protocol is known or not.

In retrospect, I probably should have commented on the goebish code I am using (I defaulted to this one because of the previous comments). 

FYI, I randomly tried an install of the multi code on the Uno, but I hit a validation error because the Uno is not listed as a supported board (but it probably would not work anyway, because I’m not using a 4-in-1 module?).

I believe the hardware to work, because if I skip the blue PCB binding, the code sends responses from the module back as strings. However, I’ll do as you say and set it to listen. In general, it’s been a lack of feedback that has made this hard to troubleshoot on my own (again, also combined with my lack of technical knowledge on the subject).

**Reply by pascallanger** [reply]
> FYI, I randomly tried an install of the multi code on the Uno, but I hit a validation error because the Uno is not listed as a supported board (but it probably would not work anyway, because I’m not using a 4-in-1 module?).

The multi documentation on this github is all about using an Arduino Nano which is why but the code can run without a single modification on an Uno since both cards are in fact using an Atmega328. Just skip the validation... In the configuration you have to indicate which RF components are installed, in your case only the nrf and which protocols to compile CX10.

> I believe the hardware to work, because if I skip the blue PCB binding, the code sends responses from the module back as strings.

No nothing here is showing me that the nrf talks with the atmega. It just means that the Arduino is connected to the PC and the program to send information on serial is working. It doesn't mean that you have connected the NRF correctly to the Uno... It will run exactly the same if you don't connect the nrf...
Have you tried just simple things like writting the nrf address register and read it back to see if you can communicate with it?

> In general, it’s been a lack of feedback that has made this hard to troubleshoot on my own.

We are talking here of things over the air so there is nothing to check per say. As I said there are just too many unknowns in your current situation. Get rid of some of the unknowns and we will be able to progress.
Check:
1. Check that you can write the nrf address register and read it back to see if it matches -> that will tell if the Uno is able to communicate with the NRF.
1. Configure the NRF to listen for bind packets to see if they match the CX10 format. You have everything in the current code to do that since the NRF does listen for an answer of the quad, just block it in listen mode and check if it ever receives something with the original TX on. -> That will really tell us if the nrf is operational and also if the CX10 packets are what we expect.

This is far away from the purpose of this github. If you want something all baked in this is the wrong place... You have to check things by yourself and get back with evidences of your tests. For now I haven't seen anything apart that you have successfuly plugged the Arduino Uno to your PC using USB, uploaded a firmware and the serial monitor is correctly set.

**Reply by dataphile-jeff** [reply]
I really appreciate your help, especially (as you said) this task is removed from the main purpose of the code.  I'm a bit busy with work, but as soon as I get a chance I am going to try each of the things you mention.

**Reply by dataphile-jeff** [reply]
@pascallanger It was a busy week, but I accomplished half of what you suggested. Specifically, I was able to write and read a different RF channel to the address register.  Also, I ran the GetStarted example for RF24 (including prettyPrintDetails) and it seems that everything is working. As you suspected, I had it wired wrong.  I don't know how the original project ever worked, because several websites indicate that MOSI, MISO, and SCK must be wired to specific pins on the Arduino, and the original project had them wired elsewhere.

I also was able to upload the full multiprotocol code to the Arduino, once I commented out the validation for the board, and selected a Custom TX endpoint.

I've been trying to check your second suggestion but I'm hitting a knowledge gap.  I've tried to advance my knowledge of the subject, but I can't figure out how to 'block it in listen mode.'  In theory, I guess I could use the RF GetStarted example to listen to the correct RF channel (which I think is 2)?  

Because the Multiprotocol can accept serial inputs from a TX, it seems tantalizing close to just cut out the part for CX10s, and then send the right commands by PC serial port.  I tried to see if I could rework the CX10_nrf24l01.ino file, which is what the original project must have done. However, the file is so intricately intertwined with the other files (Multiprotocol, Multiprotocol.h, etc.) that I'm struggling to make it work alone. 

Again, I apologize for bringing up something that is tangential to your project.  It may be that my knowledge gap on the subject is too great to make my desired project work.  If there is any advice you can give I appreciate it, but again, I want to stress that I understand if this is beyond the scope of help. In the worse case, I can hope to directly replicate Ralung's approach whenever a 4N1 module eventually arrives.
